irondb_flatbuf/irondb_flatbuf.c

```diff
@@ -0,0 +1,206 @@
+
+struct irondb_flatbuf_modstate {
+    PyObject *error_type;
+};
+
+#if PY_MAJOR_VERSION >= 3
+#define GETSTATE(m) ((struct irondb_flatbuf_modstate*)PyModule_GetState(m))
+#else
+#define GETSTATE(m) (&_state)
+static struct irondb_flatbuf_modstate _state;
+#endif
+
+static PyObject * metric_find_results(PyObject *m, PyObject *args) {
+    struct irondb_flatbuf_modstate *st = GETSTATE(m);
+    char *buffer;
+    int buffer_len;
+    if (!PyArg_ParseTuple(args, "s#", &buffer, &buffer_len))
+        return NULL;
+    // TODO FIXME
+    // Copy data for memory alignment for flatcc
+    // Python data is not memory aligned!
+    char *aligned_buffer = (char *)malloc(buffer_len);
+    memcpy(aligned_buffer, buffer, buffer_len);
+
+    int ret = metrics_ns(MetricSearchResultList_verify_as_root(aligned_buffer, buffer_len));
+    if (ret != 0) {
+        return PyErr_Format(st->error_type,
+            "Failed to verify MetricSearchResultList: %s",
+            flatcc_verify_error_string(ret));
+    }
+
+    metrics_ns(MetricSearchResultList_table_t) metric_list = metrics_ns(MetricSearchResultList_as_root(aligned_buffer));
+    metrics_ns(MetricSearchResult_vec_t) vec = metrics_ns(MetricSearchResultList_results(metric_list));
+    size_t num_results = metrics_ns(MetricSearchResult_vec_len(vec));
+
+    PyObject *array = PyList_New(num_results);
+    for (size_t i = 0; i < num_results; i++) {
+        metrics_ns(MetricSearchResult_table_t) m = metrics_ns(MetricSearchResult_vec_at(vec, i));
+        PyObject *entry = PyDict_New();
+        SET_METRIC_STR(entry, m, MetricSearchResult, name);
+
+        int leaf = metrics_ns(MetricSearchResult_leaf(m));
+        PyDict_SetItem(entry, PyUnicode_FromString("leaf"), PyBool_FromLong(leaf));
+
+        if (leaf && metrics_ns(MetricSearchResult_leaf_data_is_present(m))) {
+            metrics_ns(LeafData_table_t) ld = metrics_ns(MetricSearchResult_leaf_data(m));
+            PyObject *leaf_dict = PyDict_New();
+
+            SET_METRIC_STR(leaf_dict, ld, LeafData, uuid);
+            SET_METRIC_STR(leaf_dict, ld, LeafData, check_name);
+            PyDict_SetItem(leaf_dict, PyUnicode_FromString("name"),
+                PyUnicode_FromString(metrics_ns(LeafData_metric_name(ld))));
+            SET_METRIC_STR(leaf_dict, ld, LeafData, category);
+            SET_METRIC_STR(leaf_dict, ld, LeafData, egress_function);
+
+            PyDict_SetItem(entry, PyUnicode_FromString("leaf_data"), leaf_dict);
+        }
+        PyList_SET_ITEM(array, i, entry);
+    }
+    free(aligned_buffer);
+    return array;
+}
+
+static PyObject * metric_get_results(PyObject *m, PyObject *args) {
+    struct irondb_flatbuf_modstate *st = GETSTATE(m);
+    char *buffer;
+    int buffer_len;
+    if (!PyArg_ParseTuple(args, "s#", &buffer, &buffer_len))
+        return NULL;
+    // TODO FIXME
+    // Copy data for memory alignment for flatcc
+    // Python data is not memory aligned!
+    char *aligned_buffer = (char *)malloc(buffer_len);
```

Comment:
This should be not necessary.. I wonder if there are python buffer apis to fix this alignment problem